sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -246,85 +260,238 @@ object SpecifiedWindowFrame {
-  override lazy val resolved = false
-}
+  def this(input: Expression, offset: Expression) = this(input, offset, Literal(null))
 
-case class WindowExpression(
-    windowFunction: WindowFunction,
-    windowSpec: WindowSpecDefinition) extends Expression with Unevaluable {
+  def this(input: Expression) = this(input, Literal(1))
 
-  override def children: Seq[Expression] = windowFunction :: windowSpec :: Nil
+  def this() = this(Literal(null))
 
-  override def dataType: DataType = windowFunction.dataType
-  override def foldable: Boolean = windowFunction.foldable
-  override def nullable: Boolean = windowFunction.nullable
+  val offsetSign = -1
+}
 
-  override def toString: String = s"$windowFunction $windowSpec"
+abstract class AggregateWindowFunction extends DeclarativeAggregate with WindowFunction {
+  self: Product =>
+  override val frame = SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow)
+  override def dataType: DataType = IntegerType
+  override def nullable: Boolean = false
+  override val mergeExpressions = Nil // TODO how to deal with this?
+}
+
+abstract class RowNumberLike extends AggregateWindowFunction {
+  override def children: Seq[Expression] = Nil
+  override def inputTypes: Seq[AbstractDataType] = Nil
+  protected val zero = Literal(0)
+  protected val one = Literal(1)
+  protected val rowNumber = AttributeReference("rowNumber", IntegerType, nullable = false)()
+  override val aggBufferAttributes: Seq[AttributeReference] = rowNumber :: Nil
+  override val initialValues: Seq[Expression] = zero :: Nil
+  override val updateExpressions: Seq[Expression] = Add(rowNumber, one) :: Nil
 }
 
 /**
- * Extractor for making working with frame boundaries easier.
+ * A [[SizeBasedWindowFunction]] needs the size of the current window for its calculation.
  */
-object FrameBoundaryExtractor {
-  def unapply(boundary: FrameBoundary): Option[Int] = boundary match {
-    case CurrentRow => Some(0)
-    case ValuePreceding(offset) => Some(-offset)
-    case ValueFollowing(offset) => Some(offset)
-    case _ => None
+trait SizeBasedWindowFunction extends AggregateWindowFunction {
+  protected def n: AttributeReference = SizeBasedWindowFunction.n
+}
+
+object SizeBasedWindowFunction {
+  val n = AttributeReference("window__partition__size", IntegerType, nullable = false)()
+}
+
+case class RowNumber() extends RowNumberLike {
+  override val evaluateExpression = Cast(rowNumber, IntegerType)
+}
+
+case class CumeDist() extends RowNumberLike with SizeBasedWindowFunction {
+  override def dataType: DataType = DoubleType
+  override val frame = SpecifiedWindowFrame(RangeFrame, UnboundedPreceding, CurrentRow)
+  override val evaluateExpression = Divide(Cast(rowNumber, DoubleType), Cast(n, DoubleType))
+}
+
+case class NTile(buckets: Expression) extends RowNumberLike with SizeBasedWindowFunction {
+  def this() = this(Literal(1))
+
+  // Validate buckets.
+  buckets.eval() match {
```

Comment:
Good point. Yes and No. The ```buckets``` value only has to be constant within a partition, it would also work if the value is part of the partitioning clause. It is - however - quite a bit of work to get that in. For now I'd rather enforce a global constant number of buckets. What do you think?